src/charm.py

```diff
@@ -35,83 +28,83 @@ def __init__(self, *args: typing.Any):
+        self.framework.observe(self.on.get_ingress_data_action, self._on_get_ingress_data_action)
+
+    def _reconcile(self, _: ops.EventBase) -> None:
+        """Handle changed configuration."""
+        if not self.unit.is_leader():
+            self.unit.status = ops.BlockedStatus(
+                "this charm only supports a single unit, please remove the additional units "
+                f"using `juju scale-application {self.app.name} 1`"
+            )
+            return
+        self.unit.status = ops.ActiveStatus()
+        domain = self.config.get("domain")
+        tunnel_token = self._get_tunnel_tokens()
+        if not (domain and tunnel_token):
+            missing = []
+            if not domain:
+                missing.append("domain")
+            if not tunnel_token:
+                missing.append("tunnel-token")
+            self.unit.status = ops.BlockedStatus(f"waiting for {', '.join(missing)} configuration")
+            self._unpublish_ingress_url()
+            return
+        if relation := self.model.get_relation("cloudflared-route"):
+            self._cloudflare_route.set_tunnel_token(tunnel_token, relation=relation)
+            for ingress_relation in self._ingress.relations:
+                self._ingress.publish_url(ingress_relation, f"https://{domain}")
+        else:
+            self._unpublish_ingress_url()
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+    def _unpublish_ingress_url(self) -> None:
+        """Unpublish ingress url."""
+        for relation in self._ingress.relations:
+            self._ingress.wipe_ingress_data(relation)
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+    def _get_tunnel_tokens(self) -> str | None:
+        """Receive tunnel tokens from charm configuration.
 
-        Args:
-            event: event triggering the handler.
+        Returns:
+            Cloudflared tunnel token.
         """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ops.ActiveStatus()
-
-    def _on_config_changed(self, event: ops.ConfigChangedEvent) -> None:
-        """Handle changed configuration.
+        secret_id = typing.cast(str, self.config.get("tunnel-token"))
+        if secret_id:
+            secret = self.model.get_secret(id=secret_id)
+            return secret.get_content(refresh=True)["tunnel-token"]
```

Comment:
Any error handling required here to avoid getting in ErrorStatus in case of badly formatted secret ?

Reply:
probably set to Blocked status in that case

Reply:
Updated to put the charm in blocked state, thanks!